File: quickparcel/quickparcel/src/main/java/edu/cit/lariosa/quickparcel/entity/Delivery.java
package edu.cit.lariosa.quickparcel.entity;

import jakarta.persistence.*;
import lombok.Data;
import java.time.LocalDateTime;

@Entity
@Table(name = "deliveries")
@Data
public class Delivery {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "tracking_number", unique = true, nullable = false)
    private String trackingNumber;

    @ManyToOne
    @JoinColumn(name = "sender_id", nullable = false)
    private Sender sender;

    @ManyToOne
    @JoinColumn(name = "rider_id")
    private Rider rider;

    @OneToOne
    @JoinColumn(name = "parcel_id", unique = true, nullable = false)
    private Parcel parcel;

    @Column(name = "pickup_address", nullable = false)
    private String pickupAddress;

    @Column(name = "dropoff_address", nullable = false)
    private String dropoffAddress;

    @Column(name = "pickup_latitude")
    private Double pickupLatitude;

    @Column(name = "pickup_longitude")
    private Double pickupLongitude;

    @Column(name = "dropoff_latitude")
    private Double dropoffLatitude;

    @Column(name = "dropoff_longitude")
    private Double dropoffLongitude;

    private Double distance;

    @Column(name = "estimated_cost")
    private Double estimatedCost;

    @Column(name = "actual_cost")
    private Double actualCost;

    @Column(nullable = false)
    private String status; // PENDING, ACCEPTED, PICKED_UP, IN_TRANSIT, DELIVERED, CANCELLED

    @Column(name = "scheduled_time")
    private LocalDateTime scheduledTime;

    @Column(name = "picked_up_time")
    private LocalDateTime pickedUpTime;

    @Column(name = "delivered_time")
    private LocalDateTime deliveredTime;

    private String notes;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (trackingNumber == null) {
            trackingNumber = "QP-" + System.currentTimeMillis();
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: mobile/app/src/main/java/com/quickparcel/app/shared/models/Delivery.kt
package com.quickparcel.app.shared.models

data class Delivery(
    val id: Int,
    val trackingNumber: String,
    val pickupAddress: String,
    val dropoffAddress: String,
    val status: String,
    val estimatedCost: Double,
    val distance: Double? = null,
    val paymentMethod: String? = null,
    val paymentStatus: String? = null,
    val parcel: Parcel? = null,
    val rider: Rider? = null,
    val sender: Sender? = null,
    val createdAt: String,
    val updatedAt: String? = null,
    val deliveredTime: String? = null,
    val notes: String? = null
)

data class Parcel(
    val id: Int,
    val name: String,
    val description: String? = null,
    val weight: Double,
    val size: String,
    val category: String,
    val isFragile: Boolean = false,
    val sender: Sender? = null
)

data class Rider(
    val userId: Int,
    val user: User? = null
)

data class Sender(
    val userId: Int,
    val user: User? = null
)

data class User(
    val id: Int,
    val email: String,
    val firstName: String,
    val lastName: String,
    val phone: String? = null,
    val userType: String,
    val isActive: Boolean = true,
    val createdAt: String? = null
)